### PR TITLE
Makefile: remove support for building protobuf, C++ libraries and executables

### PR DESCRIPTION
**Pull request by jtattermusch**
More makefile simplification.

Because Makefile is now only used for building wrapped language artifacts (everything else is built either with bazel or cmake), we have no need for building C++ libraries (`libgrpc++` and others) in the Makefile. Since we have no need for C++ libraries, we can also remove support for building protobuf (which itself adds quite a bit of complexity). And since we don't need protobuf & protoc, we can drop support for building executable binaries (we only need to build libraries with Makefile).

### CODE REVIEW

**Comment by jtattermusch** [comment]
make is currently used for building protoc artifacts: as part of this PR, we need to switch protoc artifact build to cmake (and do that without reducing the backwards compatibility or those artifacts).

**Comment by jtattermusch** [comment]
Adhoc distribtest: https://g3c.corp.google.com/results/invocations/139225c4-b873-442b-bc8a-59cf7b0b3edb/targets

**Comment by jtattermusch** [comment]
@veblush  this is now ready for review.

**Comment by jtattermusch** [comment]
Problems to resolve:
Android build failure:
https://source.cloud.google.com/results/invocations/3fcde016-9a4b-4864-ae8f-7ee237fc53cc/log

Ruby build failure on Linux and Mac
https://source.cloud.google.com/results/invocations/da3c9c6f-8e33-495f-af02-98399ea46f3c/targets/github%2Fgrpc%2Ftoplevel_run_tests_invocations%2Frun_tests_ruby_linux_dbg_native/tests
https://source.cloud.google.com/results/invocations/01f3adaa-75c1-4f78-a861-25d6d0f7d453/targets/github%2Fgrpc%2Ftoplevel_run_tests_invocations%2Frun_tests_ruby_macos_dbg_native/tests

**Review by veblush — APPROVED** [review]

**Comment by drfloob** [comment]
Will this pose a problem if we introduce a C++-based I/O framework? For example, custom I/O functions were written in C to support Python's needs in iomgr, but that code will need to be ported to work as a custom class instead.

**Comment by drfloob** [comment]
Chatting offline, it seems as if this won't be a problem. This PR removes the Makefiles' ability to build the C++ libraries, which wrapped languages don't use anyhow.

**Comment by jtattermusch** [comment]
> Chatting offline, it seems as if this won't be a problem. This PR removes the Makefiles' ability to build the C++ libraries, which wrapped languages don't use anyhow.

Yes this PR is very unlikely to negatively influence your work (in fact it might simplify it by reducing complexity of our build systems). This PR only removes C++ libraries such as gRPC C++ from the Makefile, it doesn't affect the way C-core is built (or what's contained in it). Btw all the targets removed by this PR will stay buildable with cmake and bazel - we are just removing redundant targets from the Makefile.

**Comment by jtattermusch** [comment]
Bloat diff needs to be updated: https://github.com/grpc/grpc/blob/600dd1b86c1c0b4cf4bafd21391f15f4b44b0738/tools/profiling/bloat/bloat_diff.py#L81

**Comment by jtattermusch** [comment]
I also needed to update a few other builds:
- ruby codegen tests now use cmake to build protoc and grpc_ruby_plugin
- grpc_android.sh also uses cmake to build protoc and grpc_cpp_plugin (this is actually better since the rest of the android native build also uses cmake, so this actually leads to more consistency).
- bloat_diff.py now uses cmake to build grpc and grpc++ shared libraries.

**Comment by jtattermusch** [comment]
The ruby native gem build timeout is b/174656729

**Comment by jtattermusch** [comment]
@apolcyn  feel free to review the ruby changes.

**Comment by jtattermusch** [comment]
adhoc distribtests (just to make sure): https://g3c.corp.google.com/results/invocations/20613d14-ec8f-4dce-a6f2-aed6ae065def/targets

File: templates/tools/dockerfile/interoptest/grpc_interop_ruby/Dockerfile.template [hunk 1]
```diff
@@ -14,11 +14,12 @@
   # See the License for the specific language governing permissions and
   # limitations under the License.
   
-  <%include file="../../debian_jessie_header.include"/>
+  FROM debian:buster
   
   <%include file="../../apt_get_basic.include"/>
   <%include file="../../python_deps.include"/>
   <%include file="../../ruby_deps.include"/>
+  <%include file="../../cmake.include"/>
```
Comment (apolcyn) [inline templates/tools/dockerfile/interoptest/grpc_interop_ruby/Dockerfile.template:22]: why do we need to include cmake dependencies in this docker image?

Note that this docker image is used for interop tests only, and the ruby test that's changed here (which now builds protoc with cmake), doesn't run in this docker image

File: tools/run_tests/helper_scripts/build_ruby.sh [hunk 2]
```diff
@@ -25,4 +25,12 @@ rm -rf ./tmp
 rake compile
 
 # build grpc_ruby_plugin
-make grpc_ruby_plugin -j8
+mkdir -p cmake/build
+pushd cmake/build
+cmake -DgRPC_BUILD_TESTS=OFF -DCMAKE_BUILD_TYPE=Release ../..
```
Comment (apolcyn) [inline tools/run_tests/helper_scripts/build_ruby.sh:30]: this should build in Debug mode, if `GRPC_CONFIG=dbg`

File: templates/tools/dockerfile/interoptest/grpc_interop_ruby/Dockerfile.template (hunk 1 above)
Comment (jtattermusch) [reply templates/tools/dockerfile/interoptest/grpc_interop_ruby/Dockerfile.template]: removed cmake from ruby interop docker image.

**Comment by jtattermusch** [comment]
Tests are looking good, all review comments addressed. Merging.

**Comment by stanley-cheung** [comment]
This seems to have broken `grpc_xds_ruby`: https://fusion.corp.google.com/projectanalysis/summary/KOKORO/prod:grpc%2Fcore%2Fmaster%2Flinux%2Fgrpc_xds_ruby.

I think the Ruby xDS tests were relying on the `ruby_jessie_x64` image, which was renamed to `ruby_buster_x64`. So may have to fix this line: https://github.com/grpc/grpc/blob/master/tools/internal_ci/linux/grpc_xds_ruby.sh#L23.

**Comment by stanley-cheung** [comment]
The `grpc_xds_ruby` job is back to green now after #24994 is merged.

**Pull request merged**